plugins/view-transitions/js/view-transitions.js

```diff
@@ -0,0 +1,186 @@
+	const getViewTransitionEntries = ( bodyElement, articleElement ) => {
+		const globalEntries = Object.entries(
+			config.globalTransitionNames || {}
+		).map( ( [ selector, name ] ) => {
+			const element = bodyElement.querySelector( selector );
+			return [ element, name ];
+		} );
+
+		const postEntries = articleElement
+			? Object.entries( config.postTransitionNames || {} ).map(
+					( [ selector, name ] ) => {
+						const element =
+							articleElement.querySelector( selector );
+						return [ element, name ];
+					}
+			  )
+			: [];
+
+		return [ ...globalEntries, ...postEntries ];
+	};
+
+	/**
+	 * Temporarily sets view transition names for the given entries until the view transition has been completed.
+	 *
+	 * @param {Array[]}       entries   View transition entries as received from `getViewTransitionEntries()`.
+	 * @param {Promise<void>} vtPromise Promise that resolves after the view transition has been completed.
+	 * @return {Promise<void>} Promise that resolves after the view transition names were reset.
+	 */
+	const setTemporaryViewTransitionNames = async ( entries, vtPromise ) => {
+		for ( const [ element, name ] of entries ) {
+			if ( ! element ) {
+				continue;
+			}
+			element.style.viewTransitionName = name;
+		}
+
+		await vtPromise;
+
+		for ( const [ element ] of entries ) {
+			if ( ! element ) {
+				continue;
+			}
+			element.style.viewTransitionName = '';
+		}
+	};
+
+	/**
+	 * Appends a selector to another selector.
+	 *
+	 * This supports selectors which technically include multiple selectors (separated by comma).
+	 *
+	 * @param {string} selectors Main selector.
+	 * @param {string} append    Selector to append to the main selector.
+	 * @return {string} Combined selector.
+	 */
+	const appendSelectors = ( selectors, append ) => {
+		return selectors
+			.split( ',' )
+			.map( ( subselector ) => subselector.trim() + ' ' + append )
+			.join( ',' );
+	};
+
+	/**
+	 * Gets a post element (the first on the page, in case there are multiple).
+	 *
+	 * @return {Element|null} Post element, or null if none is found.
+	 */
+	const getArticle = () => {
+		if ( ! config.postSelector ) {
+			return null;
+		}
+		return document.querySelector( config.postSelector );
+	};
+
+	/**
+	 * Gets the post element for a specific post URL.
+	 *
+	 * @param {string} url Post URL (permalink) to find post element.
+	 * @return {Element|null} Post element, or null if none is found.
+	 */
+	const getArticleForUrl = ( url ) => {
+		if ( ! config.postSelector ) {
+			return null;
+		}
+		const postLinkSelector = appendSelectors(
+			config.postSelector,
+			'a[href="' + url + '"]'
+		);
+		const articleLink = document.querySelector( postLinkSelector );
+		if ( ! articleLink ) {
+			return null;
+		}
+		return articleLink.closest( config.postSelector );
+	};
+
+	/**
+	 * Customizes view transition behavior on the URL that is being navigated from.
+	 *
+	 * @param {Event} event Event fired as the previous URL is about to unload.
```

Comment:
```suggestion
	 * @param {PageSwapEvent} event Event fired as the previous URL is about to unload.
```

That should help with those errors below.

Reply:
I had it like that originally actually, but that would give me other errors about `PageSwapEvent` / `PageRevealEvent` being undefined.

Reply:
I made several updates in https://github.com/WordPress/performance/pull/1999/commits/8545b4fd820d96cb97e6227c316f4e5364f1933d, which resolves all errors except those related to `PageSwapEvent` and `PageRevealEvent`.

I am confused why these warnings still appear because:
* Both `PageSwapEvent` and `PageRevealEvent` are declared in TypeScript's own `dom` library. I verified that they're there in the version we currently use.
* The errors all refer to `Event` (instead of the above two), which I'm puzzled why they do that, even though the JS and TS code refers to `PageSwapEvent` and `PageRevealEvent`.

Let me know if you have any ideas.

Reply:
Hmm looking at 8545b4fd820d96cb97e6227c316f4e5364f1933d, instead of adding such a `ExtendedWindow` type, you should extend the `Window` interface. Like so:

```ts
declare global {
	interface Window {
		plvtInitViewTransitions?: InitViewTransitionsFunction;
		navigation?: {
			activation: NavigationActivation;
		};
	}
}
```

There's no need for this `const win = window;` part (and it doesn't affect minification anyway, so the comment there is incorrect).

The `PageSwapEvent` errors I can look at tomorrow.

Reply:
Great, thanks, fixed in https://github.com/WordPress/performance/pull/1999/commits/eaf90601f19c690ac93da192ff944ff3cda488a4. I took the comment from Optimization Detective's `detect.js`, which assigns `const win = window` with that comment, so I thought that had a purpose.

Reply:
Regarding the events, thanks to AI I fixed the TS errors in https://github.com/WordPress/performance/pull/1999/commits/3ca288e6adf5f3ac0edc470ecce917193aaa8c23, it was simply a missing type annotation in JS on the parameter. It's a bit confusing because I thought it would look at the function `@type` annotation or the parameter's `@param` type, but I guess not. Adding a `@type` annotation directly inline with the `event` parameter fixed the errors.

What remains are the ESLint errors about `PageSwapEvent` and `PageRevealEvent` being undefined. For now I disabled this check in https://github.com/WordPress/performance/pull/1999/commits/ec2904e185861cf05b2a545a9a95a72cccc53c7b for this plugin specifically, because that code is now obviously correct and it's probably just some ESLint config issue that we have to address. But of course ideally we still keep the check active and fix it. Not sure how though.

Reply:
In that case I would follow https://github.com/WordPress/performance/pull/1999#discussion_r2071541187 for now

Reply:
> Great, thanks, fixed in [eaf9060](https://github.com/WordPress/performance/commit/eaf90601f19c690ac93da192ff944ff3cda488a4). I took the comment from Optimization Detective's `detect.js`, which assigns `const win = window` with that comment, so I thought that had a purpose.

For this line:

https://github.com/WordPress/performance/blob/27592d144e60cea3bbe0eff3e27509f1937552ac/plugins/optimization-detective/detect.js#L34

The purpose was indeed to facilitate minification, since the `window` global symbol won't be used in favor of `win` which can be further reduced during minification to something like `w`.

Reply:
AFAIK minifiers automatically do that kind of stuff. I'd be very surprised if one needs to help them like that 🤔

Reply:
I just tried eliminating `win` in favor of reusing `window` and the minification used in this repo doesn't introduce any minified alias:

```js
if (0 === window.innerWidth || 0 === window.innerHeight) return void y("Window must have non-zero dimensions for URL Metric collection.");
```

Otherwise, when minified it is passing through the symbol as-is:

```js
if (0 === win.innerWidth || 0 === win.innerHeight) return void y("Window must have non-zero dimensions for URL Metric collection.");
```

I'm also surprised that the existing minifier isn't doing something smarter by reducing `win` to `w` since it is a local variable in the module.

Reply:
Regarding the ESLint problem, I applied the suggested code from @ShyamGadde, that's indeed better. Everything passes now 👍